scripts/generate_mnx_shortlists.py

```diff
@@ -0,0 +1,109 @@
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# MetaNetX has been published as:
+# MetaNetX/MNXref - reconciliation of metabolites and biochemical reactions to
+# bring together genome-scale metabolic networks
+# Sébastien Moretti, Olivier Martin, T. Van Du Tran, Alan Bridge,
+# Anne Morgat and Marco Pagni
+# Nucleic Acids Research (2016), 44(D1):D523-D526
+# https://doi.org/10.1093/bioinformatics/btt036
+# https://www.metanetx.org/
+#
+
+"""Scripts to generate a reduced shortlist from MetaNetX database dumps."""
+
+import pandas as pd
+
+# Read the database dumps in the correct format: remove comment lines, and
+# provide appropriate column names.
+
+xref = pd.read_csv(
+    "chem_xref.tsv",
+    delim_whitespace=True,
+    comment='#',
+    names=['XREF', 'MNX_ID', 'Evidence', 'Description'])
+
+
+def xref_splitter(xref):
+    """
+    Return database name from a combined string like "bigg" in "bigg:pyr".
+
+    Return 'mnx' as the ID when the string in xref contains 'MNX' but does not
+    contain a colon.
+
+    Parameters
+    ----------
+    xref : pandas.Dataframe
+        The metabolic model under investigation.
+
+    """
+    if ':' in xref:
+        return xref.split(':')[0]
+    elif 'MNX' in xref and ':' not in xref:
+        return 'mnx'
+
+
+# In chem_xref.tsv, the xref entries are key value pairs separated by ':'
+# Such as: bigg:10fthf
+# Split xref strings in MNX dump to obtain database names.
+xref['XREF_ID'] = xref['XREF'].apply(xref_splitter)
+
+# Obtain the corresponding database IDs from the combined strings in MNX dump.
+# This step includes handling old MNX IDs which are stored as
+# deprecated:MNXM9996 and makes them available in the mapping.
+xref['XREF'] = xref['XREF'].apply(lambda x: x.split(':')[1] if ':' in x else x)
+
+# Group the data in the xref dataframe so that one MNX ID maps to all
+# corresponding cross-references from other databases. Then list all
+# identifiers that belong to these databases:
+# MNX_ID    XREF_ID
+# MNXM0     chebi       [23367, 59999]
+#           metacyc     [UNKNOWN]
+groups = xref.groupby(['MNX_ID', 'XREF_ID'])
+xref = groups.apply(lambda x: list(x['XREF']))
+
+# Make a separate column for every XREF_ID:
+# MNX_ID    chebi           metacyc
+# MNXM0     [23367, 59999]  [UNKNOWN]
+xref = xref.unstack('XREF_ID')
+
+# Transpose the dataframe such that the index are now xref databases and the
+# column names are metanetx IDs.
+xref = xref.T
+
+# This is the current shortlist, expand this and rerun the script when adding
+# new code that requires the identification of key metabolites.
+# Included in the list below are energy metabolites, redox cofactors,
+# essential amino acids and some vitamins.
+shortlist = ['MNXM3', 'MNXM7', 'MNXM51', 'MNXM30', 'MNXM63', 'MNXM220',
```

Comment:
I think that a better approach would be to store the identifiers that are currently in global variables (as done for energy couples and in other places) in YAML or CSV files. This script could then load those files and generate the shortlist from there. Seems more future proof to me.

Reply:
I'm not entire sure what you mean here and how to implement it for every single 'hardcoded' identifier. Since it is working quite well right now, I'd like to hold off on this somewhat larger change until we've published the tool. I do agree with you that handling these lists of hardcoded identifiers is important as soon as there are more than two people working on this, but I don't see this as a priority right now. I'd like to save it for the #GreatRefactoring2019 ;)

Reply:
Opened this issue for us to keep track of it:
https://github.com/opencobra/memote/issues/324